### ISSUE TITLE
How can I pass a struct to a shader?

### ISSUE THREAD

**Issue by nicokoch** [tomaka/glium#941]
If my vertex shader looks like this:
```
#version 150
in vec3 position;
uniform struct Light {
   vec3 position;
   vec3 color;
} light;

void main() {
    // code omitted
}
```
How would I pass the uniform variable 'light' to the shader with glium?

**Comment by tomaka** [comment]
Hmm, that's a problem.
Passing `uniform! { light.position: ..., light.color: ... }` should work, except that dots wouldn't be accepted by the macro.

**Comment by nicokoch** [comment]
Thanks for the reply. Maybe I will create a pull request during the next days.

Additionally, it would be nice if we could pass arrays to the shader:
```
uniform struct Light {
   vec4 position;
   vec3 color;
} allLights[MAX_LIGHTS];

```
Again, I'm not really sure how to implement this.

**Comment by tomaka** [comment]
Arrays aren't really supported for the moment. They are a bit tricky and easy to get wrong, so I haven't tackled that yet :-/

**Comment by tomaka** [comment]
I took a look at it. To sum things up, let's say you have:

```glsl
struct Light {
    vec4 foo[3];
    vec3 bar;
}

uniform Light lights[10];
uniform vec4 positions[6];
```

 - It is guaranteed that `positions[0]` to `positions[5]` have contiguous indices.
 - Each element of `lights` has no relation with the others elements.
 - `lights[n].foo` and `lights[n].bar` have no relation with one another.
 - It is guaranteed that `lights[n].foo[0]` to `foo[3]` have contiguous indices for each value of n.
 - The introspection API will report `lights[0].foo[0]`, `lights[0].bar` and `positions[0]`. I don't know if it also reports other values than `0` for `lights`.

I think that multi-dimensional arrays are treated exactly the same as single-dimensional arrays, but I'm not sure.

**Comment by sergeysherkunov** [comment]
If I understand the documentation, it is sufficient implement the tuple AsUniformValue for the structure as a temporary solution of this problem? For example:
```glsl
layout(std140) uniform Material {
    sampler2D texture;
    vec3      ambient_reflectances;
    vec3      diffuse_reflectances;
    vec3      specular_reflectances;
    float     shininess;
} material;
```
```rust
struct Material {
    texture:               CompressedSrgbTexture2d,
    ambient_reflectances:  Vec3,
    diffuse_reflectances:  Vec3,
    specular_reflectances: Vec3,
    shininess:             f32,
}

impl AsUniformValue for Material {
    fn as_uniform_value(&self) -> UniformValue {
        ...
    }

    fn matches(ty: &UniformType) -> bool {
        ...
    }
}
```

**Comment by tomaka** [comment]
@sergeysherkunov Yes, I think it works if you do this. Just add a `panic!()` in `as_uniform_value` (it will never be called) and always return true in `matches`.
However you can't use textures in uniform buffers like you're trying to do.

**Comment by snafudexter** [comment]
Does [this](https://docs.rs/glium/latest/glium/uniforms/index.html) work ? It works for blocks.
